velox/exec/Exchange.cpp

```diff
@@ -123,29 +128,39 @@ class LocalExchangeSource : public ExchangeSource {
             inputPage = nullptr;
           }
           numPages_ += pages.size();
-          int64_t ackSequence;
+          int64_t ackSequence = kNoReply;
           {
             std::vector<ContinuePromise> promises;
             {
               std::lock_guard<std::mutex> l(queue_->mutex());
               requestPending_ = false;
+              uint64_t bytes = 0;
               for (auto& page : pages) {
+                bytes += page->size();
                 queue_->enqueueLocked(std::move(page), promises);
               }
+              queue_->recordReplyLocked(bytes);
               if (atEnd) {
                 queue_->enqueueLocked(nullptr, promises);
                 atEnd_ = true;
               }
-              ackSequence = sequence_ = sequence + pages.size();
+              if (pages.size() > 0) {
+                ackSequence = sequence_ = sequence + pages.size();
+              }
             }
             for (auto& promise : promises) {
               promise.setValue();
             }
           }
-          // Outside of queue mutex.
+          // Outside of queue mutex. Sends a delete results, ack or new
+          // request to self and requests more sources if space in the queue.
+          if (queue_->requestIfDue(self, ackSequence)) {
```

Comment:
We discussed that a cleaner API would to be not pass 'self' or 'ackSequence' to the query, but rather just as whether 'this' should keep trying to fetch more data.